internal/sourceanalysis/rust.go

```diff
@@ -0,0 +1,275 @@
+package sourceanalysis
+
+import (
+	"bytes"
+	"debug/dwarf"
+	"debug/elf"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/google/osv-scanner/internal/cachedregexp"
+	"github.com/google/osv-scanner/internal/sourceanalysis/ar"
+	"github.com/google/osv-scanner/pkg/models"
+	"github.com/google/osv-scanner/pkg/reporter"
+	"github.com/ianlancetaylor/demangle"
+)
+
+const RustFlagsEnv = "RUSTFLAGS=-C opt-level=3 -C debuginfo=1"
+const RustLibExtension = ".rcgu.o/"
+
+func rustAnalysis(r reporter.Reporter, pkgs []models.PackageVulns, source models.SourceInfo) {
+	binaryPaths, err := rustBuildSource(r, source)
+	if err != nil {
+		r.PrintError(fmt.Sprintf("failed to build cargo/rust project from source: %s", err))
+		return
+	}
+
+	isCalledMap := map[string]bool{}
+
+	for _, path := range binaryPaths {
+		if strings.HasSuffix(path, ".rlib") {
+			// Is a library, so need an extra step to extract the object binary file before passing to parseDWARFData
+			objFilePath, err := extractRlibArchive(path)
+			if err != nil {
+				r.PrintError(fmt.Sprintf("failed to analyse '%s': %s", path, err))
+				continue
+			}
+			// TODO: Do we need to care about error on deletion here?
+			defer os.Remove(objFilePath)
+		}
+		calls, err := parseDWARFData(r, path)
+		if err != nil {
+			r.PrintError(fmt.Sprintf("failed to analyse '%s': %s", path, err))
+			continue
+		}
+
+		for _, pv := range pkgs {
+			for _, v := range pv.Vulnerabilities {
+				for _, a := range v.Affected {
+					// Example of RUSTSEC function level information:
+					//
+					// "affects": {
+					//     "os": [],
+					//     "functions": [
+					//         "smallvec::SmallVec::grow"
+					//     ],
+					//     "arch": []
+					// }
+					ecosystemAffects, ok := a.EcosystemSpecific["affects"].(map[string]interface{})
+					if !ok {
+						continue
+					}
+					affectedFunctions, ok := ecosystemAffects["functions"].([]interface{})
+					if !ok {
+						continue
+					}
+					for _, f := range affectedFunctions {
+						if funcName, ok := f.(string); ok {
+							_, called := calls[funcName]
+							// Once one advisory marks this vuln as called, always mark as called
+							isCalledMap[v.ID] = isCalledMap[v.ID] || called
```

Comment:
nit: can we just track this as a `map[string]struct{}`, and just set this unconditionally without boolean logic make things slightly simpler?

Reply:
Hmm I think that will work, but I feel that's more confusing than the current version? The current way clearly shows 3 states in `isCalledMap`:
- There is function level vuln info, but it **wasn't** called   (true)
- There is function level vuln info, and it **is** called    (false)
- There is **no** functional level vuln info, so we don't know whether it is called (doesn't exist)

Reply:
Added some additional comments and changed variable names to make this clearer.